jmclient/jmclient/wallet_service.py

```diff
@@ -0,0 +1,362 @@
+
+    def add_restart_callback(self, callback):
+        """ Sets the function that will be
+        called in the event that the wallet
+        sync completes with a restart warning.
+        The only argument is a message string,
+        which the calling function displays to
+        the user before quitting gracefully.
+        """
+        self.restart_callback = callback
+
+    def request_sync_wallet(self):
+        """ Ensures wallet sync is complete
+        before the main event loop starts.
+        """
+        d = task.deferLater(reactor, 0.0, self.sync_wallet)
+        d.addCallback(self.start_wallet_monitoring)
+
+    def sync_wallet(self, fast=True):
+        """ Syncs wallet using fast sync.
+        TODO: enable slow sync as option here.
+        Before starting the event loop, we cache
+        the current most recent transactions as
+        reported by the blockchain interface, since
+        we are interested in deltas.
+        """
+        self.bci.sync_wallet(self.wallet, fast=fast,
+                             restart_cb=self.restart_callback)
+        self.synced = self.bci.wallet_synced
+        # Don't attempt updates on transactions that existed
+        # before startup
+        self.old_txs = self.bci.list_transactions(100)
+        return self.synced
+
+    def register_callbacks(self, callbacks, txinfo, cb_type="all"):
+        """ Register callbacks that will be called by the
+        transaction monitor loop, on transactions stored under
+        our wallet label (see WalletService.get_wallet_name()).
+        Callback arguments are currently (txd, txid) and return
+        is boolean, except "confirmed" callbacks which have
+        arguments (txd, txid, confirmations).
+        Note that callbacks MUST correctly return True if they
+        recognized the transaction and processed it, and False
+        if not. The True return value will be used to remove
+        the callback from the list.
+        Arguments:
+        `callbacks` - a list of functions with signature as above
+        and return type boolean.
+        `txinfo` - either a txid expected for the transaction, if
+        known, or a tuple of the ordered output set, of the form
+        (('script': script), ('value': value), ..). This can be
+        constructed from jmbitcoin.deserialize output, key "outs",
+        using tuple(). See WalletService.transaction_monitor().
+        `cb_type` - must be one of "all", "unconfirmed", "confirmed";
+        the first type will be called back once for every new
+        transaction, the second only once when the number of
+        confirmations is 0, and the third only once when the number
+        of confirmations is > 0.
+        """
+        if cb_type == "all":
+            # note that in this case, txid is ignored.
+            self.callbacks["all"].extend(callbacks)
+        elif cb_type in ["unconfirmed", "confirmed"]:
+            if txinfo not in self.callbacks[cb_type].keys():
```

Comment:
`.keys()` is not needed

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/7637865a7f987e40f80a22209a5efd40ba6f34b5